grease/src/Grease/Macaw/Overrides/Address.hs

```diff
@@ -0,0 +1,198 @@
+
+-- | Parse and register address overrides in the Macaw S-expression syntax.
+--
+-- Address overrides cannot use @extern@.
+loadAddressOverrides ::
+  forall arch.
+  ( Symbolic.SymArchConstraints arch
+  , ?parserHooks :: CSyn.ParserHooks (Symbolic.MacawExt arch)
+  ) =>
+  C.TypeRepr (Symbolic.ArchRegStruct arch) ->
+  C.HandleAllocator ->
+  MC.Memory (MC.ArchAddrWidth arch) ->
+  [(Integer, FilePath)] ->
+  IO (AddressOverrides arch)
+loadAddressOverrides archRegsType halloc memory paths =
+  fmap (AddressOverrides . Map.fromList) $ Monad.forM paths $ \(intAddr, path) -> do
+    segOff <-
+      case MM.resolveAbsoluteAddr memory (fromIntegral intAddr) of
+        -- TODO improve error
+        Nothing -> X.throw (GreaseException ("Bad address: " <> tshow intAddr))
+        Just segOff -> pure segOff
+    prog <- parseProgram halloc path
+    CSyn.assertNoExterns (CSyn.parsedProgExterns prog)
+    let userErr :: Text.Text -> IO a
+        userErr msg = X.throw (GreaseException (msg <> " at " <> Text.pack path))
+    case CSyn.parsedProgCFGs prog of
+      [] -> userErr "No CFGs in address override"
+      (LCCR.AnyCFG defCfg : []) -> do
+        Refl <-
+          case testEquality (LCCR.cfgArgTypes defCfg) (Ctx.Empty Ctx.:> archRegsType) of
+            Nothing -> userErr "Bad address override args"
+            Just r -> pure r
+        Refl <-
+          case testEquality (LCCR.cfgReturnType defCfg) C.UnitRepr of
+            Nothing -> userErr "Bad address override return value"
+            Just r -> pure r
+        pure (segOff, toSSA defCfg)
+      -- TODO(lb): Why does GHC 9.8.4 reject `userErr` here?
+      (_ : _) -> X.throw (GreaseException ("Too many CFGs at " <> Text.pack path))
+
+---------------------------------------------------------------------
+
+toInitialState ::
+  sym ~ W4.ExprBuilder t st fs =>
+  C.CrucibleState p sym ext rtp blocks r args ->
+  C.TypeRepr ret ->
+  C.ExecCont p sym ext (C.RegEntry sym ret) (C.OverrideLang ret) ('Just Ctx.EmptyCtx) ->
+  IO (C.ExecState p sym ext (C.RegEntry sym ret))
+toInitialState crucState retTy action = do
+  let ctx = crucState Lens.^. C.stateContext
+  C.withBackend ctx $ \_bak ->
+    pure $
+      C.InitialState
+        ctx
+        (crucState Lens.^. C.stateGlobals)
+        C.defaultAbortHandler -- (crucState Lens.^. C.abortHandler)
+        retTy
+        action
+
+-- When invoked, address overrides don't run in the current symbolic simulator.
+-- Instead, they construct a new backend and simulator, and run there. This is
+-- because Crucible can't just start running a CFG from an arbitrary state.
+runAddressOverride ::
+  ( sym ~ W4.ExprBuilder t st fs
+  , Symbolic.SymArchConstraints arch
+  ) =>
+  C.CrucibleState p sym (Symbolic.MacawExt arch) rtp blocks r (Ctx.EmptyCtx Ctx.::> Symbolic.ArchRegStruct arch) ->
+  C.SomeCFG (Symbolic.MacawExt arch) (Ctx.EmptyCtx Ctx.::> Symbolic.ArchRegStruct arch) C.UnitType ->
+  IO ()
+runAddressOverride crucState someCfg = do
```

Comment:
So if im reading this right we construct a new simstate and just pass the regs in as an arg, I guess this means we can write assertions over reg state but not memory state unfortunately, probably need an issue for that at some point since that may become important?

Reply:
We should also be able to write assertions over memory state. The LLVM memory model is a Crucible global variable, and the globals are copied as-is into the new simulator state in `toInitialState`: https://github.com/GaloisInc/grease/pull/317/files#diff-af0cd58b814678c7303457fc970cc51a0ba492caa89d1a71a03459600e38498dR126